Add an option to throw NotFoundHttpException

iamluc opened the pull request:
…instead of AccessDeniedException.

For security reason, it is sometime better to return status code 404 when a user doesn't have access to a resource to prevent leaking private informations.

This PR adds a new option on the `@Security` annotation

```
    /**
     * @Security("is_granted('POST_SHOW', post)", notFound=true)
     */
    public function showAction(Post $post)
    {
    }
```

hhamon commented:
Can't we have a more generic `statusCode` option instead which is set to `403` by default? So that we can change it and raise a `400` or `404`. The system will basically throw a regular `HttpException`.

iamluc commented:
Hi @hhamon 
Thanks for your quick reply.

I just updated the PR to use `statusCode`. I also added a `message` option to specify the message of the exception.

hhamon commented:
Nice :)

xabbuh commented on EventListener/SecurityListener.php:
```diff
@@ -65,7 +66,12 @@ public function onKernelController(FilterControllerEvent $event)
         if (!$this->language->evaluate($configuration->getExpression(), $this->getVariables($request))) {
-            throw new AccessDeniedException(sprintf('Expression "%s" denied access.', $configuration->getExpression()));
+            if ($statusCode = $configuration->getStatusCode()) {
+                throw new HttpException($statusCode, $configuration->getMessage());
```
We need a default message in case someone only modified the status code.

fabpot replied:
I would just use `Access denied` as a default message.

iamluc commented on EventListener/SecurityListener.php:
```diff
@@ -65,7 +66,12 @@ public function onKernelController(FilterControllerEvent $event)
         if (!$this->language->evaluate($configuration->getExpression(), $this->getVariables($request))) {
-            throw new AccessDeniedException(sprintf('Expression "%s" denied access.', $configuration->getExpression()));
+            if ($statusCode = $configuration->getStatusCode()) {
+                throw new HttpException($statusCode, $configuration->getMessage());
```
Do you have a suggestion ?
I'm not sure `Expression "%s" denied access.` fits for all case

fabpot commented on EventListener/SecurityListener.php:
```diff
@@ -65,7 +66,12 @@ public function onKernelController(FilterControllerEvent $event)
+                throw new HttpException($statusCode, $configuration->getMessage());
+            }
+
+            $message = $configuration->getMessage() ?: sprintf('Expression "%s" denied access.', $configuration->getExpression());
```
could be inlined.

fabpot replied:
I would never use `Expression "%s" denied access.` as a message, but the more generic `Access denied` one.

fabpot commented on Resources/doc/annotations/security.rst:
```diff
@@ -54,6 +54,26 @@ Here is another example, making use of multiple functions in the expression::
     {
     }
 
+You can throw an exception ``Symfony\Component\HttpKernel\Exception\HttpException``
```
You can throw an HttpException exception instead of

fabpot commented on Resources/doc/annotations/security.rst:
```diff
@@ -54,6 +54,26 @@ Here is another example, making use of multiple functions in the expression::
+    {
+    }
+
+Option ``message`` will allow you to specify the message of the exception::
```
allows you to

fabpot commented:
I cannot push on this PR, so I'm going to close it and submit a new one with some tweaks.

fabpot commented:
see #510
